examples/llm_plus_gnn/glem.py

```diff
@@ -0,0 +1,372 @@
+    # text_val_loader = DataLoader(valid_dataset, batch_size=lm_batch_size * 4,
+    #                              drop_last=False, pin_memory=True,
+    #                              shuffle=False)
+    text_data_loader = DataLoader(text_dataset, batch_size=lm_batch_size * 4,
+                                  drop_last=False, pin_memory=False,
+                                  shuffle=False)
+    print('done')
+
+    # =========================== GNN Data Loader =============================
+    initial_memory = torch.cuda.memory_allocated()
+    data = data.to(device)
+
+    current_memory_1 = torch.cuda.max_memory_allocated()
+    gpu_usage = float(current_memory_1 - initial_memory) / 1073741824
+    # Print the maximum memory usage after running the model
+    print(f'GPU memory usage -- data to gpu: {gpu_usage:.2f} GB')
+
+    print('build GNN dataloader(GraphSAGE NeighborLoader)', end='-->')
+    from torch_geometric.loader import NeighborLoader
+
+    # train on gold data w/o pseudo labels
+    graph_pretrain_loader = NeighborLoader(
+        data,
+        input_nodes=gold_idx,
+        num_neighbors=[15, 10, 5],
+        batch_size=gnn_batch_size,
+        shuffle=True,
+        num_workers=12,
+        persistent_workers=True,
+    )
+
+    # graph data loader w/ pseudo labels in M-step
+    graph_train_loader = NeighborLoader(
+        data,
+        input_nodes=train_idx,
+        num_neighbors=[15, 10, 5],
+        batch_size=gnn_batch_size,
+        shuffle=True,
+        num_workers=12,
+        persistent_workers=True,
+    )
+
+    # for gnn inference
+    subgraph_loader = NeighborLoader(
+        data,
+        input_nodes=None,
+        num_neighbors=[-1],
+        batch_size=gnn_batch_size * 4,
+        num_workers=12,
+        persistent_workers=True,
+    )
+    # =========================== internal function ===========================
+
+    from ogb.nodeproppred import Evaluator
+    evaluator = Evaluator(name=f'ogbn-{dataset_name}')
+
+    def evaluate(out, split):
+        y_true = data.y.cpu()
+        y_pred = out.argmax(dim=-1, keepdim=True)
+        train_acc, val_acc, test_acc = None, None, None
+        if 'train' in split:
+            train_acc = evaluator.eval({
+                'y_true': y_true[split_idx['train']],
+                'y_pred': y_pred[split_idx['train']],
+            })['acc']
+        if 'valid' in split:
+            val_acc = evaluator.eval({
+                'y_true': y_true[split_idx['valid']],
+                'y_pred': y_pred[split_idx['valid']],
+            })['acc']
+        if 'test' in split:
+            test_acc = evaluator.eval({
+                'y_true': y_true[split_idx['test']],
+                'y_pred': y_pred[split_idx['test']],
+            })['acc']
+
+        return train_acc, val_acc, test_acc
+
+    # =========================== Build GNN Model =============================
+    gnn = None
+    if args.gnn_model == 'SAGE':
```

Comment:
this should be controlled by the argparser (hidden channels, num layers etc, make those defaults)